include/simulator.h

```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
 
 class Device {
 public:
-  enum DeviceType {
-    DEVICE_GPU,
-    DEVICE_CPU,
-    DEVICE_COMM,
-  };
-  Device(DeviceType type, int node_id, int gpu_id, size_t capacity);
-  Device(DeviceType type, float bandwidth);
+    enum DeviceType {
+        DEVICE_COMP,
+        DEVICE_MEM,
+        DEVICE_COMM,
+    };
+    Device(std::string name, DeviceType type, int node_id, int socket_id, int device_id);
+    std::string name;
+    DeviceType type;
+    int node_id;
+    int socket_id;
+    int device_id;
+};
+
+class Comp_device : public Device {
 public:
-  int node_id, gpu_id;
-  float bandwidth;
-  size_t capacity;
-  DeviceType type;
+    enum CompDevType {
+        LOC_PROC,   //CPU
+        TOC_PROC,   //GPU
+    };
+    CompDevType comp_type;
+    size_t capacity;
+    Comp_device(std::string name, CompDevType comp_type, int node_id, int socket_id, int device_id);
+};
+
+class Mem_device : public Device {
+public:
+    enum MemDevType {
+        SYSTEM_MEM,     // DRAM on a single node
+        Z_COPY_MEM,     // Zero-copy memory betweeen CPU DRAM and all GPUs on a single node
+        GPU_FB_MEM,     // GPU framebuffer memory for a single GPU
+    };
+    MemDevType mem_type;
+    size_t capacity;
+    Mem_device(std::string name, MemDevType mem_type, int node_id, int socket_id, int device_id, size_t capacity);
+};
+
+class Comm_device : public Device {
+public:
+    enum CommDevType {
+        MEMBUS_COMM,
+        UPI_IN_COMM,
+        UPI_OUT_COMM,
+        NIC_IN_COMM,
+        NIC_OUT_COMM,
+        PCI_TO_HOST_COMM,
+        PCI_TO_DEV_COMM,
+        NVLINK_COMM,
+    };
+    CommDevType comm_type;
+    float latency;
+    float bandwidth;
+    Comm_device(std::string name, CommDevType comm_type, int node_id, int socket_id, int device_id, float latency, float bandwidth);
+};
+
+class MachineModel {
+public:
+  virtual int get_version() = 0;
+  virtual Comp_device *get_gpu(int device_id) = 0;
+  virtual Mem_device *get_gpu_fb_mem(int devicd_id) = 0;
+  virtual int get_num_gpus() = 0;
+  virtual float get_intra_node_gpu_bandwidth() = 0;
+  virtual float get_inter_node_gpu_bandwidth() = 0;
+  virtual std::vector<Comm_device *> get_comm_path(Mem_device *src_mem, Mem_device *tar_mem) = 0;
+  virtual std::string to_string() = 0;
+  int version;
+};
+
+class MachineModel_old : public MachineModel {
+public:
+  MachineModel_old(int num_nodes, int num_gpus_per_node, size_t capacity);
```

Comment:
Feels like a lot of these methods could/should be `const`

Reply:
fixed